pkg/reconcilermanager/controllers/rootsync_controller.go

```diff
@@ -779,52 +794,147 @@ func (r *RootSyncReconciler) validateRootSecret(ctx context.Context, rootSync *v
-	// we delete any ClusterRoleBinding that uses the old name.
-	// This ensures smooth migrations for users upgrading past that version boundary.
-	if err := r.deleteClusterRoleBinding(ctx, fmt.Sprintf("%s:%s", core.RootReconcilerPrefix, configsync.RootSyncName)); err != nil && !apierrors.IsNotFound(err) {
-		return errors.Wrap(err, "deleting old binding")
+func (r *RootSyncReconciler) listCurrentRoleRefs(ctx context.Context, reconcilerRef types.NamespacedName) (map[v1beta1.RootSyncRoleRef]client.Object, error) {
+	currentRoleMap := make(map[v1beta1.RootSyncRoleRef]client.Object)
+	opts := []client.ListOption{
+		r.roleBindingLabelSelector(reconcilerRef),
 	}
-
-	if roleConfig.Disabled {
-		return errors.Wrap( // wrap returns nil if err is nil
-			r.deleteClusterRoleBinding(ctx, RootSyncPermissionsName(reconcilerRef.Name)),
-			"ensuring clusterrolebinding does not exist",
-		)
+	rbList := rbacv1.RoleBindingList{}
+	if err := r.client.List(ctx, &rbList, opts...); err != nil {
+		return nil, errors.Wrap(err, "listing RoleBindings")
+	}
+	for idx, rb := range rbList.Items {
+		roleRef := v1beta1.RootSyncRoleRef{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: rb.RoleRef.Kind,
+				Name: rb.RoleRef.Name,
+			},
+			Namespace: rb.Namespace,
+		}
+		currentRoleMap[roleRef] = &rbList.Items[idx]
+	}
+	crbList := rbacv1.ClusterRoleBindingList{}
+	if err := r.client.List(ctx, &crbList, opts...); err != nil {
+		return nil, errors.Wrap(err, "listing ClusterRoleBindings")
+	}
+	for idx, crb := range crbList.Items {
+		roleRef := v1beta1.RootSyncRoleRef{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: crb.RoleRef.Kind,
+				Name: crb.RoleRef.Name,
+			},
+		}
+		currentRoleMap[roleRef] = &crbList.Items[idx]
 	}
+	return currentRoleMap, nil
+}
 
-	if roleConfig.Name == "" {
-		roleConfig.Name = "cluster-admin"
+func (r *RootSyncReconciler) cleanRBACBindings(ctx context.Context, reconcilerRef types.NamespacedName) error {
+	currentRefMap, err := r.listCurrentRoleRefs(ctx, reconcilerRef)
+	if err != nil {
+		return err
+	}
+	for _, binding := range currentRefMap {
+		if err := r.cleanup(ctx, binding); err != nil {
+			return errors.Wrap(err, "deleting RBAC Binding")
+		}
+	}
+	if err := r.deleteSharedClusterRoleBinding(ctx, RootSyncLegacyClusterRoleBindingName, reconcilerRef); err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrap(err, "deleting legacy binding")
 	}
+	if err := r.deleteSharedClusterRoleBinding(ctx, RootSyncBaseClusterRoleBindingName, reconcilerRef); err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrap(err, "deleting base binding")
+	}
+	return nil
+}
 
-	_, err := r.upsertClusterRoleBinding(ctx, reconcilerRef, roleConfig.Name)
-	return errors.Wrap(err, "upserting cluster role binding")
+func (r *RootSyncReconciler) configureClusterRoleBinding(ctx context.Context, reconcilerRef types.NamespacedName, roleRefs []v1beta1.RootSyncRoleRef) error {
+	currentRefMap, err := r.listCurrentRoleRefs(ctx, reconcilerRef)
```

Comment:
My initial thoughts is that it's fair game to assume that a resource key that is prefixed with our `GroupName` (`configsync.gke.io`) is a resource that we can manage. The same argument could be made for other resources as well, say if a user tried to manage the `configsync.gke.io/ns-reconciler` base RoleBinding in the source repo.

If this is a concern, we could potentially add another label to make it more explicit - say something like `configsync.gke.io/managed-by: reconciler-manager`.


As for tracking the inventory of bindings in RSync annotations - I'm not sure how this would scale to larger numbers of bindings. We would need to track `Kind`, `Name`, and `Namespace` for an arbitrary number of bindings. This is unlikely to fit in a single annotation, and it's not clear to me how we would shard this across multiple annotations.

I could potentially see adding new Status fields on RootSync/RepoSync, however I'm not sure how compelling the case for this is over using a label selector.